src/status_im/protocol/handlers.cljs

```diff
@@ -313,84 +289,76 @@
+                         :update-keys            {:dispatch [:update-keys-received message]}
+                         :online                 {:dispatch [:contact-online-received message]} 
+                         nil)]
+          (when (nil? route-fx) (debug "Unknown message type" type))
           (cache/add! processed-message)
           (merge
             {::save-processed-messages processed-message}
-            (when route-event {:dispatch route-event})))))))
-
-(defn update-message-status [db {:keys [message-id ack-of-message group-id from status]}]
-  (let [message-id'          (or ack-of-message message-id) 
-        update-group-status? (and group-id (not= status :sent))
-        message-path         [:chats (or group-id from) :messages message-id']
-        current-status       (if update-group-status?
-                               (get-in db (into message-path [:user-statuses from :status]))
-                               (get-in db (into message-path [:message-status])))]
-    ;; for some strange reason, we sometimes receive status update for message we don't have,
-    ;; that's why the first condition in if
-    (if (and (get-in db message-path)
-             (not= :seen current-status))
-      (if update-group-status?
-        (assoc-in db (into message-path [:user-statuses from]) {:whisper-identity from
-                                                                :status           status})
-        (assoc-in db (into message-path [:message-status]) status))
-      db)))
+            route-fx))))))
 
 (handlers/register-handler-fx
   :update-message-status
   [re-frame/trim-v
+   (re-frame/inject-cofx :get-stored-message)
    (re-frame/inject-cofx ::chats-is-active?)]
-  (fn [{db :db chats-is-active? :chats-is-active?}
-       [{:keys                                        [from]
-         {:keys [message-id ack-of-message group-id]} :payload
-         :as message}
-        status]]
-    (let [data {:status status
-                :message-id message-id :ack-of-message ack-of-message
-                :group-id group-id :from from}]
-      (merge
-        {::save-message-status! data}
-        (when (= status :delivered)
-         {:dispatch  [:pending-message-remove message]})
-        (when chats-is-active?
-          {:db (update-message-status db data)})))))
+  (fn [{:keys [db chats-is-active? get-stored-message]} [{:keys [from sent-from payload]} status]]
+    (let [message-identifier (get-message-id payload)
+          message-db-path    [:chats (or (:group-id payload) from) :messages message-identifier] 
+          from-id            (or sent-from from) 
+          message            (get-stored-message message-identifier)]
+      ;; proceed with updating status if chat is active, message is in realm and existing
+      ;; status is not already "final" = `:seen` by the user sending the update
+      (when (and chats-is-active? message (not= :seen (get message :user-statuses from-id)))
```

Comment:
@janherich can you extract `message (not= :seen (get message :user-statuses from-id))` so that in result condition will be
```clojure
(when (and chats-is-active? (not (seen? message from-id))) ..
```